### PR TITLE
rustdoc: redirect URLs

### PR DESCRIPTION
**Pull request by nrc**
cc #35020 which does this properly

r? @alexcrichton

### CODE REVIEW

**Comment by nrc** [comment]
cc @steveklabnik, @ollie27

**Comment by alexcrichton** [comment]
@nrc so just to clarify, this emits `Foo.v.html` which redirects to `struct.Foo.html`?

**Comment by ollie27** [comment]
As with the previous PR, methods and struct fields are in the same namespace so we get conflicts. For example `std::ops::Range` has two `start.v` ids. Is that not a problem?

**Comment by nrc** [comment]
> As with the previous PR, methods and struct fields are in the same namespace so we get conflicts. For example std::ops::Range has two start.v ids. Is that not a problem?

It is, but without changing the canonical URLs/anchors there doesn't seem to be a great way to disambiguate. However, since these are not canonical URLs, it doesn't really matter - for tool users who link to Rustdoc, there will be a very few URLs that don't work properly, but I expect this to be a pretty small edge case.

**Comment by nrc** [comment]
> so just to clarify, this emits Foo.v.html which redirects to struct.Foo.html?

Well, `Foo.t.html`, but yes that is basically what is happening here, plus the extra anchors for 'sub-items'.

**Comment by ollie27** [comment]
There are a few issues with the implementation:

* Nested `<a>` tags are invalid so you'll need to use another tag for the new ids.
* ids are supposed to be unique so you'll need to call `derive_id` for all the new ones.
* There should probably be a few new tests 😉.

**Comment by alexcrichton** [comment]
@bors: r-

er... maybe a mis-parse?

**Comment by nrc** [comment]
bors is clearly hungry for this PR, it won't take 'no' for an answer

**Comment by nrc** [comment]
re tests, does anyone (@ollie27?) have suggestions for how to test the existence of the redirects? I don't see a way to fit it into the existing framework, but I don't know it very well.

**Comment by alexcrichton** [comment]
I think you can at least test for their existence with the rustdoc tests, and maybe html tags inside them as well?

**Comment by ollie27** [comment]
Check out [test/rustdoc/redirect.rs](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/redirect.rs) and [test/rustdoc/issue-33302.rs](https://github.com/rust-lang/rust/blob/master/src/test/rustdoc/issue-33302.rs) for examples which test the same kind of things.

For example I'd do something like the following for structs:

```rust
#![crate_name = "foo"]

// @has 'foo/FooStruct.t.html'
// @has - '//p/a' 'struct.FooStruct.html'
// @has 'foo/struct.FooStruct.html'
pub struct FooStruct {
    // @has - '//*[@id="foo.v"]' 'foo'
    pub foo: (),
    // @has - '//*[@id="dupe.v"]' 'dupe'
    pub dupe: (),
}
impl FooStruct {
    // @has - '//*[@id="bar.v"]' 'bar'
    pub fn bar() {}
    // @has - '//*[@id="dupe.v-1"]' 'dupe'
    pub fn dupe() {}
}
```

You should be able repeat that for every other item type.

**Comment by nrc** [comment]
@ollie27 I tried something like this and it doesn't work. I'm not really clear on the semantics of `@has`, but I'm assuming it is ensuring that the text string occurs in the generated document. So, your example is saying that there should be a link to each of `struct.FooStruct.html` and `FooStruct.t.html`, which there isn't - there is only a link to the former. Is my understanding correct? Is there a way to check for the existence of a file?

**Comment by ollie27** [comment]
Check out [src/etc/htmldocck.py](https://github.com/rust-lang/rust/blob/master/src/etc/htmldocck.py) for a better explanation than I could give.

**Comment by nrc** [comment]
Thanks for the link, I knew these must be documented somewhere.

> `PATH` is relative to the output directory. It can be given as `-` which repeats the most recently used `PATH`.

This is what was tripping me up.

**Comment by nrc** [comment]
Updated with tests, spans instead of anchors, and unique ids.

@alexcrichton r?

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -2268,9 +2286,12 @@ fn item_struct(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
         if fields.peek().is_some() {
             write!(w, "<h2 class='fields'>Fields</h2>")?;
             for (field, ty) in fields {
-                write!(w, "<span id='{shortty}.{name}' class='{shortty}'><code>{name}: {ty}</code>
-                           </span><span class='stab {stab}'></span>",
-                       shortty = ItemType::StructField,
+                write!(w, "<span id='{item_type}.{name}' class='{item_type}'>
+                           <span id='{name}.{name_space}' class='invisible'>
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: You'll need to use `derive_id` for these as well. The existing ids are also missing a call to `derive_id` so you may as well also fix that while you're at it.

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -2339,8 +2360,10 @@ fn item_enum(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
     if !e.variants.is_empty() {
         write!(w, "<h2 class='variants'>Variants</h2>\n")?;
         for variant in &e.variants {
-            write!(w, "<span id='{shortty}.{name}' class='variant'><code>{name}",
-                   shortty = ItemType::Variant,
+            write!(w, "<span id='{item_type}.{name}' class='variant'>\
+                       <span id='{name}.{name_space}' class='invisible'><code>{name}",
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: Same here.

File: src/librustdoc/html/render.rs [hunk 3]
```diff
@@ -2366,9 +2389,12 @@ fn item_enum(w: &mut fmt::Formatter, cx: &Context, it: &clean::Item,
                     if let StructFieldItem(ref ty) = field.inner {
                         write!(w, "<tr><td \
                                    id='variant.{v}.field.{f}'>\
-                                   <code>{f}:&nbsp;{t}</code></td><td>",
+                                   <span id='{v}.{vns}.{f}.{fns}' class='invisible'>\
```
Comment (ollie27) [inline src/librustdoc/html/render.rs]: Same here.

**Comment by ollie27** [comment]
I was hoping you'd add a new test covering every single item type. For example enum variant struct fields and macros are different so need explicit tests.

Are you going to document this feature anywhere? What kind of stability guarantees do these new URLs have? Can other people use them?

**Comment by nrc** [comment]
> I was hoping you'd add a new test covering every single item type. For example enum variant struct fields and macros are different so need explicit tests.

I suppose a few more tests couldn't hurt. The main code paths are exercised and the coverage seems as good or better than existing Rustdoc tests.

> Are you going to document this feature anywhere?

No. I see no point in encouraging use of these until the long-term URL solution is figured out.

> What kind of stability guarantees do these new URLs have?

None, use at your own risk.

> Can other people use them?

Sure, but I wouldn't encourage it.

> You'll need to use derive_id for these as well. The existing ids are also missing a call to derive_id so you may as well also fix that while you're at it.

I was assuming there was some logic behind not calling derive_id here (e.g., you can't have duplicate named fields, so I assume there can't be a name clash).

**Comment by ollie27** [comment]
> The main code paths are exercised and the coverage seems as good or better than existing Rustdoc tests.

The test coverage in rustdoc isn't very good right now so we should be aiming for better coverage.

> I was assuming there was some logic behind not calling derive_id here (e.g., you can't have duplicate named fields, so I assume there can't be a name clash).

`derive_id` is also used by the markdown so I was worried it might conflict there but having a closer look, it turns out the markdown will never create an id containing a `.` so technically they can't conflict. It wouldn't harm to call `derive_id` though, we definitely need to call it for the new ones anyway.

**Pull request merged**